rust/kernel/file_operations.rs

```diff
@@ -249,6 +301,70 @@ macro_rules! declare_file_operations {
     };
 }
 
+/// Allows the handling of ioctls defined with the `_IO`, `_IOR`, `_IOW`, and `_IOWR` macros.
+///
+/// For each macro, there is a handler function that takes the appropriate types as arguments.
+pub trait IoctlHandler: Sync {
+    /// Handles ioctls defined with the `_IO` macro, that is, with no buffer as argument.
+    fn pure(&self, _file: &File, _cmd: u32, _arg: usize) -> KernelResult<i32> {
+        Err(Error::EINVAL)
+    }
+
+    /// Handles ioctls defined with the `_IOR` macro, that is, with an output buffer provided as
+    /// argument.
+    fn read(&self, _file: &File, _cmd: u32, _writer: &mut UserSlicePtrWriter) -> KernelResult<i32> {
+        Err(Error::EINVAL)
+    }
+
+    /// Handles ioctls defined with the `_IOW` macro, that is, with an input buffer provided as
+    /// argument.
+    fn write(
+        &self,
+        _file: &File,
+        _cmd: u32,
+        _reader: &mut UserSlicePtrReader,
+    ) -> KernelResult<i32> {
+        Err(Error::EINVAL)
+    }
+
+    /// Handles ioctls defined with the `_IOWR` macro, that is, with a buffer for both input and
+    /// output provided as argument.
+    fn read_write(&self, _file: &File, _cmd: u32, _data: UserSlicePtr) -> KernelResult<i32> {
+        Err(Error::EINVAL)
+    }
+}
+
+/// Dispatches the given ioctl to the appropriate handler based on the value of `cmd`. It also
+/// creates a [`UserSlicePtr`], [`UserSlicePtrReader`], or [`UserSlicePtrWriter`] depending on the
+/// direction of the buffer of the command.
+///
+/// It is meant to be used in implementations of [`FileOperations::ioctl`] and
+/// [`FileOperations::compat_ioctl`].
+pub fn dispatch_ioctl<T: IoctlHandler>(
+    handler: &T,
+    file: &File,
+    cmd: u32,
+    arg: usize,
+) -> KernelResult<i32> {
+    let dir = (cmd >> bindings::_IOC_DIRSHIFT) & bindings::_IOC_DIRMASK;
+    if dir == bindings::_IOC_NONE {
+        return T::pure(handler, file, cmd, arg);
+    }
+
+    let size = (cmd >> bindings::_IOC_SIZESHIFT) & bindings::_IOC_SIZEMASK;
+
+    // SAFETY: The caller is reponsible for setting the appropriate `fs`. Additionally, we only
+    // create one instance of the user slice, so TOCTOU issues are not possible.
+    let data = unsafe { UserSlicePtr::new(arg as _, size as _) }?;
```

Comment:
Since the function is not `unsafe`, what stops the caller providing a broken `fs` here?

Also, two nits: typo `responsible` and `fs` could be more clear what it refers to, since it is not an argument.

Reply:
The justification was that this was called from an `ioctl` callback, so the kernel must have set `fs` appropriately.

However, it is possible for this to be called from elsewhere, so I'll make this function unsafe, and push the safety requirement to the caller. In binder's case, it will be as I stated above, i.e., the kernel will set fs appropriately before calling a file's ioctl callback.

Reply:
Done.